libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -0,0 +1,77 @@
+package com.mapbox.navigation.core.routerefresh
+
+import android.util.Log
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.route.RouteRefreshCallback
+import com.mapbox.navigation.base.route.RouteRefreshError
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.utils.timer.MapboxTimer
+import java.util.concurrent.TimeUnit
+import kotlinx.coroutines.Job
+
+/**
+ * This class is responsible for refreshing the current direction route's traffic.
+ * This does not support alternative routes.
+ *
+ * If the route is successfully refreshed, this class will update the [TripSession.route]
+ *
+ * [start] and [stop] are attached to the application lifecycle. Observing routes that
+ * can be refreshed are handled by this class. Calling [start] will restart the refresh timer.
+ */
+internal class RouteRefreshController(
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession
+) {
+    private val routerRefreshTimer = MapboxTimer()
+
+    init {
+        routerRefreshTimer.restartAfterMillis = TimeUnit.SECONDS.toMillis(10)
+    }
+
+    fun start(): Job {
+        stop()
+        return routerRefreshTimer.startTimer {
+            val route = tripSession.route
+            val legIndex = tripSession.getRouteProgress()?.currentLegProgress()?.legIndex() ?: 0
+            if (supportsRefresh(route)) {
+                directionsSession.requestRouteRefresh(
+                    route!!,
```

Comment:
Should be this nullable https://github.com/mapbox/mapbox-navigation-android/blob/ee15527ea3d9ac5226ce44adc659b6ca5484e191/libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/DirectionsSession.kt#L26? How are we handling `route!!` scenario, `tripSession.route` could be `null` and it'll crash at runtime.

Reply:
`supportsRefresh(route: DirectionsRoute?)` checks if it's null.. maybe that's a little unclear

kotlin uses `null` in such a nice way, maybe it's more clear to make the function use null. updated